digi/xbee/recovery.py

```diff
@@ -184,6 +197,20 @@ def autorecover_device(self):
             if not error:
                 break
 
+        # If we received an error on all the preferred bauds that were
+        # auto-detected as likely candidates, then attempt the
+        # rest of the remaining reasonable bauds that it could be at
+        # before giving up
+        if error:
+            rest_of_bauds = [
+                b for b in _RECOVERY_ALL_REASONABLE_BAUDRATES
+                if b not in recovery_baudrate
+            ]
+            for baudrate in rest_of_bauds:
+                error = self._try_baudrate(baudrate)
```

Comment:
I am not sure about this change.
How long does it take to execute each `_try_baudrate()`? Did we try this modifications in a running gateway with an XBee that does not respond at all?
If the XBee is not responsible at all (and must be reprogrammed), this increase dramatically the time to fail and may impact in the whole gateway behavior.
Maybe we should think another way without modifying the recovery process...

Reply:
Hi @tatianaleon !
We added this change, because we have found that the auto-detection magic that is done with the _RECOVERY_CHAR_TO_BAUDRATE table doesn't always work/find the proper bauds.

For example, on my Gateway, I have my baud set to 9600, but the auto-detection doesn't seem to work properly on it:

2023-12-19 20:36:05,382 digi.xbee.recovery:    DEBUG: Autorecovering the device by entering in recovery mode
2023-12-19 20:36:10,400 digi.xbee.recovery:    DEBUG: Databytes read from recovery are '00'
2023-12-19 20:36:10,403 digi.xbee.recovery:    DEBUG: XBee baudrate is [230400, 4800, 2400, 1200]

So we felt that it might be better to try the other "common bauds" before giving up.
This is what led to this patch.

Perhaps we could add a parameter to the recovery() functions, so let us specify a deeper/harder attempt at recovery/baudrates before we give up?
Would that be OK?

PS: We are unsure of how that _RECOVERY_CHAR_TO_BAUDRATE table was derived...
How was each entry determined as to what character sequence would indicate what baud rates that it may be at?

Thanks!

Reply:
Also:

> How long does it take to execute each `_try_baudrate()`?

It should take at most about 5 seconds per incorrect baud rate:
* `enter_at_command_mode` takes up to 3 seconds: 1 second sleep, read data for up to 2 seconds looking for `OK\r`
* Back out in `_try_baudrate`, the `for command in ...` loop should bail at the first command if we're at the wrong baud rate, and it should bail after at most 2 seconds (in case of no output at all)
  * If we read _anything_ though, the `while self._xbee_serial_port.inWaiting() == 0 ...` loop will exit early.

I would be okay with us removing some of the less-common (or probably-less-common) baud rates from our list of bauds to try. For example, I doubt anyone out there is configuring an XBee at lower than 9600. Can't really comment on whether anyone uses any particular one of the higher baud rates (19200, etc).

Reply:
@skilau, what are you using to test this functionality? The [RecoverSerialConnection](https://github.com/digidotcom/xbee-python/blob/master/examples/configuration/RecoverSerialConnection/RecoverSerialConnection.py) example can help you.

I changed the list of baud rates to only have 460800.
This is working with an XBee 3 (HV: 4244, FW: 3012) configured at 1200, 2400, 4800, 9600, 19200, 38400,  57600, 115200, 230400, 921600, connected to a Linux PC.
I also tested the same example in the gateway (changing previously the baud rate to the list above) and it is also working.

```
# python /opt/recovery-xbee.py 

 +------------------------------------------------------+
 | XBee Python Library Recover serial connection Sample |
 +------------------------------------------------------+
    
Opening the XBee device by forcing its baudrate to 460800
Device opened and set to operate at 460800 bauds
#
```